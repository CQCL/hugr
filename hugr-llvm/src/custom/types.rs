use std::marker::PhantomData;

use itertools::Itertools as _;

use hugr::types::CustomType;

use anyhow::Result;
use inkwell::types::{BasicMetadataTypeEnum, BasicType as _, BasicTypeEnum, FunctionType};

pub use crate::utils::type_map::CustomTypeKey;

use crate::{
    sum::LLVMSumType,
    types::{HugrFuncType, HugrSumType, TypingSession},
    utils::type_map::TypeMapping,
};

pub trait LLVMCustomTypeFn<'a>:
    for<'c> Fn(TypingSession<'c, 'a>, &CustomType) -> Result<BasicTypeEnum<'c>> + 'a
{
}

impl<
        'a,
        F: for<'c> Fn(TypingSession<'c, 'a>, &CustomType) -> Result<BasicTypeEnum<'c>> + 'a + ?Sized,
    > LLVMCustomTypeFn<'a> for F
{
}

#[derive(Default, Clone)]
pub struct LLVMTypeMapping<'a>(PhantomData<&'a ()>);

impl<'a> TypeMapping for LLVMTypeMapping<'a> {
    type InV<'c> = TypingSession<'c, 'a>;

    type OutV<'c> = BasicTypeEnum<'c>;

    type SumOutV<'c> = LLVMSumType<'c>;

    type FuncOutV<'c> = FunctionType<'c>;

    fn sum_into_out<'c>(&self, sum: Self::SumOutV<'c>) -> Self::OutV<'c> {
        sum.as_basic_type_enum()
    }

    fn func_into_out<'c>(&self, sum: Self::FuncOutV<'c>) -> Self::OutV<'c> {
        sum.ptr_type(Default::default()).as_basic_type_enum()
    }

    fn map_sum_type<'c>(
        &self,
        sum_type: &HugrSumType,
        context: TypingSession<'c, 'a>,
        variants: impl IntoIterator<Item = Vec<Self::OutV<'c>>>,
    ) -> Result<Self::SumOutV<'c>> {
        LLVMSumType::try_new2(
            context.iw_context(),
            variants.into_iter().collect(),
            sum_type.clone(),
        )
    }

    fn map_function_type<'c>(
        &self,
        _: &HugrFuncType,
        context: TypingSession<'c, 'a>,
        inputs: impl IntoIterator<Item = Self::OutV<'c>>,
        outputs: impl IntoIterator<Item = Self::OutV<'c>>,
    ) -> Result<Self::FuncOutV<'c>> {
        let iw_context = context.iw_context();
        let inputs: Vec<BasicMetadataTypeEnum<'c>> = inputs.into_iter().map_into().collect_vec();
        let outputs = outputs.into_iter().collect_vec();
        Ok(match outputs.as_slice() {
            &[] => iw_context.void_type().fn_type(&inputs, false),
            [res] => res.fn_type(&inputs, false),
            ress => iw_context.struct_type(ress, false).fn_type(&inputs, false),
        })
    }
}
