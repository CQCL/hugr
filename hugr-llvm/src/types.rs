use std::rc::Rc;

use anyhow::Result;
use delegate::delegate;
use hugr::extension::ExtensionId;
use hugr::types::{SumType, Type, TypeName};
use inkwell::types::FunctionType;
use inkwell::{context::Context, types::BasicTypeEnum};

use crate::custom::types::{LLVMCustomTypeFn, LLVMTypeMapping};
pub use crate::sum::LLVMSumType;
use crate::utils::type_map::TypeMap;

/// A type alias for a hugr function type. We use this to disambiguate from
/// the LLVM [FunctionType].
pub type HugrFuncType = hugr::types::Signature;

/// A type alias for a hugr type. We use this to disambiguate from LLVM types.
pub type HugrType = Type;

/// A type alias for a hugr sum type.
pub type HugrSumType = SumType;

/// A type that holds [Rc] shared pointers to everything needed to convert from
/// a hugr [HugrType] to an LLVM [Type](inkwell::types).
#[derive(Clone)]
pub struct TypingSession<'c, 'a> {
    iw_context: &'c Context,
    type_converter: Rc<TypeConverter<'a>>,
}

impl<'c, 'a> TypingSession<'c, 'a> {
    delegate! {
        to self.type_converter.clone() {
            /// Convert a [HugrType] into an LLVM [Type](BasicTypeEnum).
            pub fn llvm_type(&self, [self.clone()], hugr_type: &HugrType) -> Result<BasicTypeEnum<'c>>;
            /// Convert a [HugrFuncType] into an LLVM [FunctionType].
            pub fn llvm_func_type(&self, [self.clone()], hugr_type: &HugrFuncType) -> Result<FunctionType<'c>>;
            /// Convert a hugr [HugrSumType] into an LLVM [LLVMSumType].
            pub fn llvm_sum_type(&self, [self.clone()], hugr_type: HugrSumType) -> Result<LLVMSumType<'c>>;
        }
    }

    /// Creates a new `TypingSession`.
    pub fn new(iw_context: &'c Context, type_converter: Rc<TypeConverter<'a>>) -> Self {
        Self {
            iw_context,
            type_converter,
        }
    }

    /// Returns a reference to the inner [Context].
    pub fn iw_context(&self) -> &'c Context {
        self.iw_context
    }
}

#[derive(Default)]
pub struct TypeConverter<'a>(TypeMap<'a, LLVMTypeMapping<'a>>);

impl<'a> TypeConverter<'a> {
    pub(super) fn custom_type(
        &mut self,
        custom_type: (ExtensionId, TypeName),
        handler: impl LLVMCustomTypeFn<'a>,
    ) {
        self.0.set_callback(custom_type, handler);
    }

    pub fn llvm_type<'c>(
        self: Rc<Self>,
        context: TypingSession<'c, 'a>,
        hugr_type: &HugrType,
    ) -> Result<BasicTypeEnum<'c>> {
        self.0.map_type(hugr_type, context)
    }

    pub fn llvm_func_type<'c>(
        self: Rc<Self>,
        context: TypingSession<'c, 'a>,
        hugr_type: &HugrFuncType,
    ) -> Result<FunctionType<'c>> {
        self.0.map_function_type(hugr_type, context)
    }

    pub fn llvm_sum_type<'c>(
        self: Rc<Self>,
        context: TypingSession<'c, 'a>,
        hugr_type: HugrSumType,
    ) -> Result<LLVMSumType<'c>> {
        self.0.map_sum_type(&hugr_type, context)
    }

    pub fn session<'c>(self: Rc<Self>, iw_context: &'c Context) -> TypingSession<'c, 'a> {
        TypingSession::new(iw_context, self)
    }
}

#[cfg(test)]
#[allow(drop_bounds)]
pub mod test {

    use hugr::{
        std_extensions::arithmetic::int_types::INT_TYPES,
        type_row,
        types::{SumType, Type},
    };

    use insta::assert_snapshot;
    use rstest::rstest;

    use crate::{extension::int::add_int_extensions, test::*, types::HugrFuncType};

    #[rstest]
    #[case(0,HugrFuncType::new(type_row!(Type::new_unit_sum(2)), type_row!()))]
    #[case(1, HugrFuncType::new(Type::new_unit_sum(1), Type::new_unit_sum(3)))]
    #[case(2,HugrFuncType::new(vec![], vec![Type::new_unit_sum(1), Type::new_unit_sum(1)]))]
    fn func_types(#[case] _id: i32, #[with(_id)] llvm_ctx: TestContext, #[case] ft: HugrFuncType) {
        assert_snapshot!(
            "func_type_to_llvm",
            llvm_ctx.get_typing_session().llvm_func_type(&ft).unwrap(),
            &ft.to_string()
        )
    }

    #[rstest]
    #[case(0, SumType::new_unary(0))]
    #[case(1, SumType::new_unary(1))]
    #[case(2,SumType::new([vec![Type::new_unit_sum(0), Type::new_unit_sum(1)], vec![Type::new_unit_sum(2), Type::new_unit_sum(3)]]))]
    #[case(3, SumType::new_unary(2))]
    fn sum_types(#[case] _id: i32, #[with(_id)] llvm_ctx: TestContext, #[case] st: SumType) {
        assert_snapshot!(
            "sum_type_to_llvm",
            llvm_ctx
                .get_typing_session()
                .llvm_sum_type(st.clone())
                .unwrap(),
            &st.to_string()
        )
    }

    #[rstest]
    #[case(0, INT_TYPES[0].clone())]
    #[case(1, INT_TYPES[3].clone())]
    #[case(2, INT_TYPES[4].clone())]
    #[case(3, INT_TYPES[5].clone())]
    #[case(4, INT_TYPES[6].clone())]
    #[case(5, Type::new_sum([vec![INT_TYPES[2].clone()]]))]
    #[case(6, Type::new_sum([vec![INT_TYPES[6].clone(),Type::new_unit_sum(1)], vec![Type::new_unit_sum(2), INT_TYPES[2].clone()]]))]
    #[case(7, Type::new_function(HugrFuncType::new(type_row!(Type::new_unit_sum(2)), Type::new_unit_sum(3))))]
    fn ext_types(#[case] _id: i32, #[with(_id)] mut llvm_ctx: TestContext, #[case] t: Type) {
        llvm_ctx.add_extensions(add_int_extensions);
        assert_snapshot!(
            "type_to_llvm",
            llvm_ctx.get_typing_session().llvm_type(&t).unwrap(),
            &t.to_string()
        );
    }
}
