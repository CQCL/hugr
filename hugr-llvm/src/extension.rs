pub mod collections;
pub mod conversions;
pub mod float;
pub mod int;
pub mod logic;
pub mod prelude;

pub use prelude::{DefaultPreludeCodegen, PreludeCodegen, PreludeCodegenExtension};
