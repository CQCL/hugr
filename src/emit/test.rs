use std::iter;

use crate::custom::int::add_int_extensions;
use crate::types::HugrFuncType;
use hugr::builder::DataflowSubContainer;
use hugr::builder::{
    BuildHandle, Container, DFGWrapper, Dataflow, HugrBuilder, ModuleBuilder, SubContainer,
};
use hugr::extension::prelude::BOOL_T;
use hugr::extension::{ExtensionRegistry, EMPTY_REG};
use hugr::ops::constant::CustomConst;
use hugr::ops::handle::FuncID;
use hugr::ops::{CallIndirect, Tag, UnpackTuple, Value};
use hugr::std_extensions::arithmetic::int_ops::{self, INT_OPS_REGISTRY};
use hugr::std_extensions::arithmetic::int_types::ConstInt;
use hugr::types::{Signature, Type, TypeRow};
use hugr::{type_row, Hugr};
use itertools::Itertools;
use rstest::rstest;

use crate::test::*;

#[allow(clippy::upper_case_acronyms)]
pub type DFGW<'a> = DFGWrapper<&'a mut Hugr, BuildHandle<FuncID<true>>>;

pub struct SimpleHugrConfig {
    ins: TypeRow,
    outs: TypeRow,
    extensions: ExtensionRegistry,
}

impl SimpleHugrConfig {
    pub fn new() -> Self {
        Self {
            ins: Default::default(),
            outs: Default::default(),
            extensions: EMPTY_REG,
        }
    }

    pub fn with_ins(mut self, ins: impl Into<TypeRow>) -> Self {
        self.ins = ins.into();
        self
    }

    pub fn with_outs(mut self, outs: impl Into<TypeRow>) -> Self {
        self.outs = outs.into();
        self
    }

    pub fn with_extensions(mut self, extensions: ExtensionRegistry) -> Self {
        self.extensions = extensions;
        self
    }

    pub fn finish(
        self,
        make: impl for<'a> FnOnce(DFGW<'a>) -> <DFGW<'a> as SubContainer>::ContainerHandle,
    ) -> Hugr {
        self.finish_with_exts(|builder, _| make(builder))
    }
    pub fn finish_with_exts(
        self,
        make: impl for<'a> FnOnce(
            DFGW<'a>,
            &ExtensionRegistry,
        ) -> <DFGW<'a> as SubContainer>::ContainerHandle,
    ) -> Hugr {
        let mut mod_b = ModuleBuilder::new();
        let func_b = mod_b
            .define_function("main", HugrFuncType::new(self.ins, self.outs))
            .unwrap();
        make(func_b, &self.extensions);
        mod_b.finish_hugr(&self.extensions).unwrap()
    }
}

impl Default for SimpleHugrConfig {
    fn default() -> Self {
        Self::new()
    }
}

#[macro_export]
macro_rules! check_emission {
    ($hugr: ident, $test_ctx:ident) => {
        let root = $crate::fat::FatExt::fat_root::<hugr::ops::Module>(&$hugr).unwrap();
        let module = $test_ctx
            .get_emit_hugr()
            .emit_module(root)
            .unwrap()
            .finish();

        let mut settings = insta::Settings::clone_current();
        let new_suffix = settings
            .snapshot_suffix()
            .map_or("pre-mem2reg".into(), |x| format!("pre-mem2reg@{x}"));
        settings.set_snapshot_suffix(new_suffix);
        settings.bind(|| insta::assert_snapshot!(module.to_string()));

        module
            .verify()
            .unwrap_or_else(|pp| panic!("Failed to verify module: {pp}"));

        let pb = inkwell::passes::PassManager::create(());
        pb.add_promote_memory_to_register_pass();
        pb.run_on(&module);

        insta::assert_snapshot!(module.to_string());
    };
}

#[rstest]
fn emit_hugr_make_tuple(llvm_ctx: TestContext) {
    let hugr = SimpleHugrConfig::new()
        .with_ins(vec![BOOL_T, BOOL_T])
        .with_outs(Type::new_tuple(vec![BOOL_T, BOOL_T]))
        .finish(|mut builder: DFGW| {
            let in_wires = builder.input_wires();
            let r = builder.make_tuple(in_wires).unwrap();
            builder.finish_with_outputs([r]).unwrap()
        });
    check_emission!(hugr, llvm_ctx);
}

#[rstest]
fn emit_hugr_unpack_tuple(llvm_ctx: TestContext) {
    let hugr = SimpleHugrConfig::new()
        .with_ins(Type::new_tuple(vec![BOOL_T, BOOL_T]))
        .with_outs(vec![BOOL_T, BOOL_T])
        .finish(|mut builder: DFGW| {
            let unpack = builder
                .add_dataflow_op(
                    UnpackTuple::new(vec![BOOL_T, BOOL_T].into()),
                    builder.input_wires(),
                )
                .unwrap();
            builder.finish_with_outputs(unpack.outputs()).unwrap()
        });
    check_emission!(hugr, llvm_ctx);
}

#[rstest]
fn emit_hugr_tag(llvm_ctx: TestContext) {
    let hugr = SimpleHugrConfig::new()
        .with_outs(Type::new_unit_sum(3))
        .finish(|mut builder: DFGW| {
            let tag = builder
                .add_dataflow_op(
                    Tag::new(1, vec![vec![].into(), vec![].into(), vec![].into()]),
                    builder.input_wires(),
                )
                .unwrap();
            builder.finish_with_outputs(tag.outputs()).unwrap()
        });
    check_emission!(hugr, llvm_ctx);
}

#[rstest]
fn emit_hugr_dfg(llvm_ctx: TestContext) {
    let hugr = SimpleHugrConfig::new()
        .with_ins(Type::UNIT)
        .with_outs(Type::UNIT)
        .finish(|mut builder: DFGW| {
            let dfg = {
                let b = builder
                    .dfg_builder(HugrFuncType::new_endo(Type::UNIT), builder.input_wires())
                    .unwrap();
                let w = b.input_wires();
                b.finish_with_outputs(w).unwrap()
            };
            builder.finish_with_outputs(dfg.outputs()).unwrap()
        });
    check_emission!(hugr, llvm_ctx);
}

#[rstest]
fn emit_hugr_conditional(llvm_ctx: TestContext) {
    let hugr = {
        let input_v_rows: Vec<TypeRow> = (0..3).map(Type::new_unit_sum).map_into().collect_vec();
        let output_v_rows = {
            let mut r = input_v_rows.clone();
            r.reverse();
            r
        };

        SimpleHugrConfig::new()
            .with_ins(vec![Type::new_sum(input_v_rows.clone()), Type::UNIT])
            .with_outs(vec![Type::new_sum(output_v_rows.clone()), Type::UNIT])
            .finish(|mut builder: DFGW| {
                let cond = {
                    let [sum_input, other_input] = builder.input_wires_arr();
                    let mut cond_b = builder
                        .conditional_builder(
                            (input_v_rows.clone(), sum_input),
                            [(Type::UNIT, other_input)],
                            vec![Type::new_sum(output_v_rows.clone()), Type::UNIT].into(),
                        )
                        .unwrap();
                    for i in 0..3 {
                        let mut case_b = cond_b.case_builder(i).unwrap();
                        let [case_input, other_input] = case_b.input_wires_arr();
                        let tag = case_b
                            .add_dataflow_op(Tag::new(2 - i, output_v_rows.clone()), [case_input])
                            .unwrap();
                        case_b
                            .finish_with_outputs([tag.out_wire(0), other_input])
                            .unwrap();
                    }
                    cond_b.finish_sub_container().unwrap()
                };
                let [o1, o2] = cond.outputs_arr();
                builder.finish_with_outputs([o1, o2]).unwrap()
            })
    };
    check_emission!(hugr, llvm_ctx);
}

#[rstest]
fn emit_hugr_load_constant(mut llvm_ctx: TestContext) {
    llvm_ctx.add_extensions(add_int_extensions);
    let v = Value::tuple([
        Value::unit_sum(2, 4).unwrap(),
        ConstInt::new_s(4, -24).unwrap().into(),
    ]);

    let hugr = SimpleHugrConfig::new()
        .with_outs(v.get_type())
        .with_extensions(INT_OPS_REGISTRY.to_owned())
        .finish(|mut builder: DFGW| {
            let konst = builder.add_load_value(v);
            builder.finish_with_outputs([konst]).unwrap()
        });
    check_emission!(hugr, llvm_ctx);
}

#[rstest]
fn emit_hugr_call(llvm_ctx: TestContext) {
    fn build_recursive(mod_b: &mut ModuleBuilder<Hugr>, name: &str, io: TypeRow) {
        let f_id = mod_b
            .declare(name, HugrFuncType::new_endo(io).into())
            .unwrap();
        let mut func_b = mod_b.define_declaration(&f_id).unwrap();
        let call = func_b
            .call(&f_id, &[], func_b.input_wires(), &EMPTY_REG)
            .unwrap();
        func_b.finish_with_outputs(call.outputs()).unwrap();
    }

    let mut mod_b = ModuleBuilder::new();
    build_recursive(&mut mod_b, "main_void", type_row![]);
    build_recursive(&mut mod_b, "main_unary", type_row![BOOL_T]);
    build_recursive(&mut mod_b, "main_binary", type_row![BOOL_T, BOOL_T]);
    let hugr = mod_b.finish_hugr(&EMPTY_REG).unwrap();
    check_emission!(hugr, llvm_ctx);
}

#[rstest]
fn emit_hugr_call_indirect(llvm_ctx: TestContext) {
    fn build_recursive(mod_b: &mut ModuleBuilder<Hugr>, name: &str, io: TypeRow) {
        let signature = HugrFuncType::new_endo(io);
        let f_id = mod_b.declare(name, signature.clone().into()).unwrap();
        let mut func_b = mod_b.define_declaration(&f_id).unwrap();
        let func = func_b.load_func(&f_id, &[], &EMPTY_REG).unwrap();
        let inputs = iter::once(func).chain(func_b.input_wires());
        let call_indirect = func_b
            .add_dataflow_op(CallIndirect { signature }, inputs)
            .unwrap();
        func_b.finish_with_outputs(call_indirect.outputs()).unwrap();
    }

    let mut mod_b = ModuleBuilder::new();
    build_recursive(&mut mod_b, "main_void", type_row![]);
    build_recursive(&mut mod_b, "main_unary", type_row![BOOL_T]);
    build_recursive(&mut mod_b, "main_binary", type_row![BOOL_T, BOOL_T]);
    let hugr = mod_b.finish_hugr(&EMPTY_REG).unwrap();
    check_emission!(hugr, llvm_ctx);
}

#[rstest]
fn emit_hugr_custom_op(mut llvm_ctx: TestContext) {
    llvm_ctx.add_extensions(add_int_extensions);
    let v1 = ConstInt::new_s(4, -24).unwrap();
    let v2 = ConstInt::new_s(4, 24).unwrap();

    let hugr = SimpleHugrConfig::new()
        .with_outs(v1.get_type())
        .with_extensions(INT_OPS_REGISTRY.to_owned())
        .finish_with_exts(|mut builder: DFGW, ext_reg| {
            let k1 = builder.add_load_value(v1);
            let k2 = builder.add_load_value(v2);
            let ext_op = int_ops::EXTENSION
                .instantiate_extension_op("iadd", [4.into()], ext_reg)
                .unwrap();
            let add = builder.add_dataflow_op(ext_op, [k1, k2]).unwrap();
            builder.finish_with_outputs(add.outputs()).unwrap()
        });
    check_emission!(hugr, llvm_ctx);
}

#[rstest]
fn get_external_func(llvm_ctx: TestContext) {
    let emc = llvm_ctx.get_emit_module_context();
    let func_type1 = emc.iw_context().i32_type().fn_type(&[], false);
    let func_type2 = emc.iw_context().f64_type().fn_type(&[], false);
    let foo1 = emc.get_extern_func("foo", func_type1).unwrap();
    assert_eq!(foo1.get_name().to_str().unwrap(), "foo");
    let foo2 = emc.get_extern_func("foo", func_type1).unwrap();
    assert_eq!(foo1, foo2);
    assert!(emc.get_extern_func("foo", func_type2).is_err());
}

#[rstest]
fn diverse_module_children(llvm_ctx: TestContext) {
    let hugr = {
        let mut builder = ModuleBuilder::new();
        let _ = {
            let fbuilder = builder
                .define_function("f1", HugrFuncType::new_endo(type_row![]))
                .unwrap();
            fbuilder.finish_sub_container().unwrap()
        };
        let _ = {
            let fbuilder = builder
                .define_function("f2", HugrFuncType::new_endo(type_row![]))
                .unwrap();
            fbuilder.finish_sub_container().unwrap()
        };
        let _ = builder.add_constant(Value::true_val());
        let _ = builder
            .declare("decl", HugrFuncType::new_endo(type_row![]).into())
            .unwrap();
        builder.finish_hugr(&EMPTY_REG).unwrap()
    };
    check_emission!(hugr, llvm_ctx);
}

#[rstest]
fn diverse_dfg_children(llvm_ctx: TestContext) {
    let hugr = SimpleHugrConfig::new()
        .with_outs(BOOL_T)
        .finish(|mut builder: DFGW| {
            let [r] = {
                let mut builder = builder
                    .dfg_builder(HugrFuncType::new(type_row![], BOOL_T), [])
                    .unwrap();
                let konst = builder.add_constant(Value::false_val());
                let func = {
                    let mut builder = builder
                        .define_function("scoped_func", HugrFuncType::new(type_row![], BOOL_T))
                        .unwrap();
                    let w = builder.load_const(&konst);
                    builder.finish_with_outputs([w]).unwrap()
                };
                let [r] = builder
                    .call(func.handle(), &[], [], &EMPTY_REG)
                    .unwrap()
                    .outputs_arr();
                builder.finish_with_outputs([r]).unwrap().outputs_arr()
            };
            builder.finish_with_outputs([r]).unwrap()
        });
    check_emission!(hugr, llvm_ctx);
}

#[rstest]
fn diverse_cfg_children(llvm_ctx: TestContext) {
    let hugr = SimpleHugrConfig::new()
        .with_outs(BOOL_T)
        .finish(|mut builder: DFGW| {
            let [r] = {
                let mut builder = builder.cfg_builder([], type_row![BOOL_T]).unwrap();
                let konst = builder.add_constant(Value::false_val());
                let func = {
                    let mut builder = builder
                        .define_function("scoped_func", HugrFuncType::new(type_row![], BOOL_T))
                        .unwrap();
                    let w = builder.load_const(&konst);
                    builder.finish_with_outputs([w]).unwrap()
                };
                let entry = {
                    let mut builder = builder
                        .entry_builder([type_row![]], type_row![BOOL_T])
                        .unwrap();
                    let control = builder.add_load_value(Value::unary_unit_sum());
                    let [r] = builder
                        .call(func.handle(), &[], [], &EMPTY_REG)
                        .unwrap()
                        .outputs_arr();
                    builder.finish_with_outputs(control, [r]).unwrap()
                };
                let exit = builder.exit_block();
                builder.branch(&entry, 0, &exit).unwrap();
                builder.finish_sub_container().unwrap().outputs_arr()
            };
            builder.finish_with_outputs([r]).unwrap()
        });
    check_emission!(hugr, llvm_ctx);
}

#[rstest]
fn load_function(llvm_ctx: TestContext) {
    let hugr = {
        let mut builder = ModuleBuilder::new();
        let target_sig = Signature::new_endo(type_row![]);
        let target_func = builder
            .declare("target_func", target_sig.clone().into())
            .unwrap();
        let _ = {
            let mut builder = builder
                .define_function(
                    "main",
                    Signature::new(type_row![], Type::new_function(target_sig)),
                )
                .unwrap();
            let r = builder.load_func(&target_func, &[], &EMPTY_REG).unwrap();
            builder.finish_with_outputs([r]).unwrap()
        };
        builder.finish_hugr(&EMPTY_REG).unwrap()
    };

    check_emission!(hugr, llvm_ctx);
}
