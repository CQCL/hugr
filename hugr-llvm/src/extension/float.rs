use anyhow::{anyhow, Result};
use hugr::ops::ExtensionOp;
use hugr::ops::{constant::CustomConst, Value};
use hugr::std_extensions::arithmetic::float_ops::FloatOps;
use hugr::{
    std_extensions::arithmetic::float_types::{self, ConstF64},
    HugrView,
};
use inkwell::{
    types::{BasicType, FloatType},
    values::{BasicValue, BasicValueEnum},
};

use crate::emit::emit_value;
use crate::emit::ops::{emit_custom_binary_op, emit_custom_unary_op};
use crate::emit::{func::EmitFuncContext, EmitOpArgs};

use crate::custom::CodegenExtsBuilder;

/// Emit a float comparison operation.
fn emit_fcmp<'c, H: HugrView>(
    context: &mut EmitFuncContext<'c, '_, H>,
    args: EmitOpArgs<'c, '_, ExtensionOp, H>,
    pred: inkwell::FloatPredicate,
) -> Result<()> {
    let true_val = emit_value(context, &Value::true_val())?;
    let false_val = emit_value(context, &Value::false_val())?;

    emit_custom_binary_op(context, args, |ctx, (lhs, rhs), _| {
        // get result as an i1
        let r = ctx.builder().build_float_compare(
            pred,
            lhs.into_float_value(),
            rhs.into_float_value(),
            "",
        )?;
        // convert to whatever BOOL_T is
        Ok(vec![ctx
            .builder()
            .build_select(r, true_val, false_val, "")?])
    })
}

fn emit_float_op<'c, H: HugrView>(
    context: &mut EmitFuncContext<'c, '_, H>,
    args: EmitOpArgs<'c, '_, ExtensionOp, H>,
    op: FloatOps,
) -> Result<()> {
    // We emit the float comparison variants where NaN is an absorbing value.
    // Any comparison with NaN is always false.
    #[allow(clippy::wildcard_in_or_patterns)]
    match op {
        FloatOps::feq => emit_fcmp(context, args, inkwell::FloatPredicate::OEQ),
        FloatOps::fne => emit_fcmp(context, args, inkwell::FloatPredicate::ONE),
        FloatOps::flt => emit_fcmp(context, args, inkwell::FloatPredicate::OLT),
        FloatOps::fgt => emit_fcmp(context, args, inkwell::FloatPredicate::OGT),
        FloatOps::fle => emit_fcmp(context, args, inkwell::FloatPredicate::OLE),
        FloatOps::fge => emit_fcmp(context, args, inkwell::FloatPredicate::OGE),
        FloatOps::fadd => emit_custom_binary_op(context, args, |ctx, (lhs, rhs), _| {
            Ok(vec![ctx
                .builder()
                .build_float_add(lhs.into_float_value(), rhs.into_float_value(), "")?
                .as_basic_value_enum()])
        }),
        FloatOps::fsub => emit_custom_binary_op(context, args, |ctx, (lhs, rhs), _| {
            Ok(vec![ctx
                .builder()
                .build_float_sub(lhs.into_float_value(), rhs.into_float_value(), "")?
                .as_basic_value_enum()])
        }),
        FloatOps::fneg => emit_custom_unary_op(context, args, |ctx, v, _| {
            Ok(vec![ctx
                .builder()
                .build_float_neg(v.into_float_value(), "")?
                .as_basic_value_enum()])
        }),
        FloatOps::fmul => emit_custom_binary_op(context, args, |ctx, (lhs, rhs), _| {
            Ok(vec![ctx
                .builder()
                .build_float_mul(lhs.into_float_value(), rhs.into_float_value(), "")?
                .as_basic_value_enum()])
        }),
        FloatOps::fdiv => emit_custom_binary_op(context, args, |ctx, (lhs, rhs), _| {
            Ok(vec![ctx
                .builder()
                .build_float_div(lhs.into_float_value(), rhs.into_float_value(), "")?
                .as_basic_value_enum()])
        }),
        // Missing ops, not supported by inkwell
        FloatOps::fmax
        | FloatOps::fmin
        | FloatOps::fabs
        | FloatOps::ffloor
        | FloatOps::fceil
        | FloatOps::ftostring
        | _ => {
            let name: &str = op.into();
            Err(anyhow!("FloatOpEmitter: unimplemented op: {name}"))
        }
    }
}

fn emit_constf64<'c, H: HugrView>(
    context: &mut EmitFuncContext<'c, '_, H>,
    k: &ConstF64,
) -> Result<BasicValueEnum<'c>> {
    let ty: FloatType = context.llvm_type(&k.get_type())?.try_into().unwrap();
    Ok(ty.const_float(k.value()).as_basic_value_enum())
}

pub fn add_float_extensions<'a, H: HugrView + 'a>(
    cem: CodegenExtsBuilder<'a, H>,
) -> CodegenExtsBuilder<'a, H> {
    cem.custom_type(
        (
            float_types::EXTENSION_ID,
            float_types::FLOAT64_CUSTOM_TYPE.name().clone(),
        ),
        |ts, _custom_type| Ok(ts.iw_context().f64_type().as_basic_type_enum()),
    )
    .custom_const(emit_constf64)
    .simple_extension_op::<FloatOps>(emit_float_op)
}

impl<'a, H: HugrView + 'a> CodegenExtsBuilder<'a, H> {
    pub fn add_float_extensions(self) -> Self {
        add_float_extensions(self)
    }
}

#[cfg(test)]
mod test {
    use hugr::extension::simple_op::MakeOpDef;
    use hugr::extension::SignatureFunc;
    use hugr::std_extensions::arithmetic::float_ops::{self, FloatOps};
    use hugr::types::TypeRow;
    use hugr::Hugr;
    use hugr::{
        builder::{Dataflow, DataflowSubContainer},
        std_extensions::arithmetic::{
            float_ops::FLOAT_OPS_REGISTRY,
            float_types::{ConstF64, FLOAT64_TYPE},
        },
    };
    use rstest::rstest;

    use super::add_float_extensions;
    use crate::{
        check_emission,
        emit::test::SimpleHugrConfig,
        test::{llvm_ctx, TestContext},
    };

    fn test_float_op(op: FloatOps) -> Hugr {
        let SignatureFunc::PolyFuncType(poly_sig) = op.signature() else {
            panic!("Expected PolyFuncType");
        };
        let sig = poly_sig.body();
        let inp: TypeRow = sig.input.clone().try_into().unwrap();
        let out: TypeRow = sig.output.clone().try_into().unwrap();

        SimpleHugrConfig::new()
            .with_ins(inp)
            .with_outs(out)
            .with_extensions(float_ops::FLOAT_OPS_REGISTRY.to_owned())
            .finish(|mut builder| {
                let outputs = builder
                    .add_dataflow_op(op, builder.input_wires())
                    .unwrap()
                    .outputs();
                builder.finish_with_outputs(outputs).unwrap()
            })
    }

    #[rstest]
    fn const_float(mut llvm_ctx: TestContext) {
        llvm_ctx.add_extensions(add_float_extensions);
        let hugr = SimpleHugrConfig::new()
            .with_outs(FLOAT64_TYPE)
            .with_extensions(FLOAT_OPS_REGISTRY.to_owned())
            .finish(|mut builder| {
                let c = builder.add_load_value(ConstF64::new(3.12));
                builder.finish_with_outputs([c]).unwrap()
            });
        check_emission!(hugr, llvm_ctx);
    }

    #[rstest]
    #[case::feq(FloatOps::feq)]
    #[case::fne(FloatOps::fne)]
    #[case::flt(FloatOps::flt)]
    #[case::fgt(FloatOps::fgt)]
    #[case::fle(FloatOps::fle)]
    #[case::fge(FloatOps::fge)]
    #[case::fadd(FloatOps::fadd)]
    #[case::fsub(FloatOps::fsub)]
    #[case::fneg(FloatOps::fneg)]
    #[case::fmul(FloatOps::fmul)]
    #[case::fdiv(FloatOps::fdiv)]
    fn float_operations(mut llvm_ctx: TestContext, #[case] op: FloatOps) {
        let name: &str = op.into();
        let hugr = test_float_op(op);
        llvm_ctx.add_extensions(add_float_extensions);
        check_emission!(name, hugr, llvm_ctx);
    }
}
