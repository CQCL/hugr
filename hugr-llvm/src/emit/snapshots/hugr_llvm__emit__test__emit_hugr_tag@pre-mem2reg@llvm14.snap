---
source: src/emit/test.rs
expression: module.to_string()
---
; ModuleID = 'test_context'
source_filename = "test_context"

define { i32, {}, {}, {} } @_hl.main.1() {
alloca_block:
  %"0" = alloca { i32, {}, {}, {} }, align 8
  %"4_0" = alloca { i32, {}, {}, {} }, align 8
  br label %entry_block

entry_block:                                      ; preds = %alloca_block
  store { i32, {}, {}, {} } { i32 1, {} poison, {} undef, {} poison }, { i32, {}, {}, {} }* %"4_0", align 4
  %"4_01" = load { i32, {}, {}, {} }, { i32, {}, {}, {} }* %"4_0", align 4
  store { i32, {}, {}, {} } %"4_01", { i32, {}, {}, {} }* %"0", align 4
  %"02" = load { i32, {}, {}, {} }, { i32, {}, {}, {} }* %"0", align 4
  ret { i32, {}, {}, {} } %"02"
}
